tests/test_echo.py

```diff
@@ -2,11 +2,142 @@
 # -*- coding: utf-8 -*-
 
 import unittest
-import echo
+import subprocess
+
 
 # Your test case class goes here
 class TestEcho(unittest.TestCase):
-    pass
+    def test_help(self):
+        """ Running the program without arguments should show usage. """
+
+        # Run the command `python ./echo.py -h` in a separate process, then
+        # collect its output.
+        process = subprocess.Popen(
+            ["python", "./echo.py", "-h"],
+            stdout=subprocess.PIPE)
+        stdout, _ = process.communicate()
+        stdout = stdout.decode("utf-8")
+        usage = open("./USAGE", "r").read()
+        self.assertEquals(stdout, usage)
+
+    def test_upper(self):
+        """Running the program with -u or --upper as arguments should result with
+        "upper" stored in namespace returned from parser.parse_args and that
+        "hello" will return to "HELLO" when the program is run"""
+
+        # Run the command `python ./echo.py -u hello` in a separate process,
+        # then collect its output.
+        process = subprocess.Popen(
+            ["python", "./echo.py", "-u", "hello"],
+            stdout=subprocess.PIPE)
+        args_u = process.args
+        stdout, _ = process.communicate()
+        stdout_u = stdout.decode("utf-8")
+
+        process = subprocess.Popen(
+            ["python", "./echo.py", "--upper", "hello"],
+            stdout=subprocess.PIPE)
+        args_upper = process.args
+        stdout, _ = process.communicate()
+        stdout_upper = stdout.decode("utf-8")
+
+        usage = open("./USAGEUPPER", "r").read()
+        self.assertEquals(stdout_u, usage)
```

Comment:
for these, your second argument, usage, could easily be just the string "HELLO" to match what you put into your process.  the extra files that you made for USAGEUPPER, etc seem unnecessary to me.

i like that you're checking both -u and --upper.  that's a good idea.  

it seems strange to me that you define args_u as process.args (including -u), and then check that -u is in it.  that's more of testing your test than testing your program.  in my opinion you could get rid of all the self.assertEqual("-u" in args_u, True) type statements.

Reply:
From the Readme: Write a unit test that asserts that upper get stored inside of the namespace returned from parser.parse_args when either "-u" or "--upper" arguments are passed.

It should also test that "hello" gets turned into "HELLO" when the program is run.

Reply:
Looking at this again, you're right that I was testing the test instead of the echo.py function. I think that the only way to "assert that upper gets stored inside the namespace returned from parser.parse_args" is that "hello" gets turned into "HELLO" when the program is run. We're not supposed to output anything else, and the tests can't access local variables in the tested program (which is what I thought I was doing). Refactored to remove those. 

Regarding testing string versus file, I wanted to get some muscle memory of how to compare using a file.

Reply:
My understanding is that madarp's method of calling echo.py's parser creator from within test_echo.py allows you to test what options would actually be stored based on the test's arguments.